### PR TITLE
Dynamic discovery of Web Console Public URL

### PR DESCRIPTION
**Pull request by simo5**
Supplants #19150 
Checks for what is the Webconsole Public URL from the proper config map and uses it for proxying there.

### CODE REVIEW

**Comment by simo5** [comment]
@deads2k so this PR kinda works, the test failures are expected given now we do not have redirects available right away.
PTAL and please let me know if you have any guidance on how these tests should be changed (for example TestRootRedirect).

Please also note that unlike your PR I stopped trying to update the public URL right away, after to attempts where all that it did was to tie up a lot of go routines all trying concurrently to update the URL and failing because the server was not healthy yet, causing tests to simply time out.

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	publicURL       atomic.Value
+	configMapGetter coreclientv1.ConfigMapsGetter
+	polling         time.Duration
+}
+
+func newWebConsolePublicURLAccessor(clientConfig rest.Config, stopCh <-chan struct{}) *webConsolePublicURLAccessor {
+	accessor := &webConsolePublicURLAccessor{
+		configMapGetter: coreclientv1.NewForConfigOrDie(&clientConfig),
+	}
+	go accessor.run(stopCh)
```
Comment (deads2k) [inline pkg/cmd/server/origin/webconsole_proxy.go]: Prefer a separate, explicit call to run.

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	return accessor
+}
+
+func (a *webConsolePublicURLAccessor) getPublicConsoleURL() string {
+	currValue, ok := a.publicURL.Load().(string)
+	if ok && len(currValue) > 0 {
+		return currValue
+	}
+
+	// if we aren't already set, too bad
```
Comment (deads2k) [inline pkg/cmd/server/origin/webconsole_proxy.go]: out of date comment

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	return a.updatePublicConsoleURL()
+}
+
+func (a *webConsolePublicURLAccessor) updatePublicConsoleURL() string {
+	configMap, err := a.configMapGetter.ConfigMaps("openshift-web-console").Get("webconsole-config", metav1.GetOptions{})
```
Comment (deads2k) [inline pkg/cmd/server/origin/webconsole_proxy.go]: // TODO best effort ratelimit maybe.  Something to think about.  Not needed for this pull.

File: test/integration/master_routes_test.go [hunk 4]
```diff
@@ -183,17 +188,79 @@ func TestRootRedirect(t *testing.T) {
 		t.Fatalf("Unexpected index: \ngot=%v,\n\n expected=%v,\n\ndiff=%v", got.Paths, expectedIndex, diff.ObjectDiff(expectedIndex, got.Paths))
 	}
 
-	req, err = http.NewRequest("GET", masterConfig.OAuthConfig.MasterPublicURL, nil)
-	req.Header.Set("Accept", "text/html")
-	resp, err = transport.RoundTrip(req)
+	// Create fake config map to test console redirect
```
Comment (deads2k) [inline test/integration/master_routes_test.go]: I'd accept a simple `apiVersion` literal.

File: pkg/cmd/server/origin/webconsole_proxy.go (hunk 3 above)
Comment (simo5) [reply pkg/cmd/server/origin/webconsole_proxy.go]: Ok I'll add it

File: pkg/cmd/server/origin/webconsole_proxy.go (hunk 1 above)
Comment (simo5) [reply pkg/cmd/server/origin/webconsole_proxy.go]: ok

File: pkg/cmd/server/origin/webconsole_proxy.go (hunk 2 above)
Comment (simo5) [reply pkg/cmd/server/origin/webconsole_proxy.go]: good catch

File: test/integration/master_routes_test.go [hunk 5]
```diff
@@ -183,17 +188,79 @@ func TestRootRedirect(t *testing.T) {
...
+		},
+		Data: map[string]string{
+			"webconsole-config.yaml": string(data),
+		},
+	}
+	if _, err = clusterAdminKubeClientset.Core().ConfigMaps("openshift-web-console").Create(&configMap); err != nil {
+		t.Fatalf("Unextecpted error: %v", err)
+	}
+
+	// try three times then give up
```
Comment (deads2k) [inline test/integration/master_routes_test.go]: A little weird given the live lookup.  I guess if we ever change it back.  Both method had merit.

File: test/integration/master_routes_test.go (hunk 4 above)
Comment (simo5) [reply test/integration/master_routes_test.go]: Sorry, I do not understand the context of this comment

**Comment by deads2k** [comment]
A few minor comments. It's a good starting point, but I have to hold it for the rebase which usually touches this area.

/approve
/hold

File: test/integration/master_routes_test.go (hunk 5 above)
Comment (simo5) [reply test/integration/master_routes_test.go]: I had issues getting this to work, and at some point I got it spin forever w/o any timeout, so I decided it was a better idea to limit it and terminate early

File: pkg/cmd/server/origin/webconsole_proxy.go [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	newValue := webConsoleConfig.ClusterInfo.ConsolePublicURL
+	a.publicURL.Store(newValue)
+
+	// Once we have a value we relax polling to once per minute
+	a.polling = 1 * time.Minute
+
+	return newValue
+}
+
+func (a *webConsolePublicURLAccessor) Run(stopCh <-chan struct{}) {
```
Comment (deads2k) [inline pkg/cmd/server/origin/webconsole_proxy.go]: Forgot.  Need to `defer utilruntime.HandleCrash()` or a panic will kill the entire process.

File: pkg/cmd/server/origin/master.go [hunk 7]
```diff
@@ -378,6 +372,10 @@ func (c *MasterConfig) buildHandlerChain(genericConfig *apiserver.Config) (func(
 	}
 
 	return func(apiHandler http.Handler, genericConfig *apiserver.Config) http.Handler {
+			// Machinery that let's use discover the Web Console Public URL
+			accessor := newWebConsolePublicURLAccessor(c.PrivilegedLoopbackClientConfig)
+			go accessor.Run(stopCh)
```
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: yeah, that's what I meant.

**Comment by simo5** [comment]
/hold cancel
@deads2k fixed HandleCrash, and rebased on master, PTAL

**Comment by deads2k** [comment]
Good starting point.  We can make updates later if we have to.  The hardest part to decide about is behavior before we've discovered it.

/lgtm

**Pull request merged**